File: problems/14_2d_dp/0097_interleaving_string/solution.dart
// 0097. Interleaving String
bool isInterleave(String s1, String s2, String s3) {
  final m = s1.length, n = s2.length;
  if (m + n != s3.length) return false;
  final dp = List<bool>.filled(n + 1, false);
  dp[0] = true;
  for (var j = 1; j <= n; j++) dp[j] = dp[j - 1] && s2[j - 1] == s3[j - 1];
  for (var i = 1; i <= m; i++) {
    dp[0] = dp[0] && s1[i - 1] == s3[i - 1];
    for (var j = 1; j <= n; j++) {
      dp[j] = (dp[j] && s1[i - 1] == s3[i + j - 1]) ||
              (dp[j - 1] && s2[j - 1] == s3[i + j - 1]);
    }
  }
  return dp[n];
}
void main() {
  if (isInterleave("aabcc", "dbbca", "aadbbcbcac") != true) throw 'a';
  if (isInterleave("aabcc", "dbbca", "aadbbbaccc") != false) throw 'b';
  if (isInterleave("", "", "") != true) throw 'c';
  if (isInterleave("", "abc", "abc") != true) throw 'd';
  if (isInterleave("a", "", "a") != true) throw 'e';
  if (isInterleave("a", "b", "ab") != true) throw 'f';
  if (isInterleave("a", "b", "ba") != true) throw 'g';
  print('All tests passed!');
}

File: problems/14_2d_dp/0097_interleaving_string/solution.py
"""
Problem: 0097. Interleaving String
"""


class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        m, n = len(s1), len(s2)
        if m + n != len(s3):
            return False
        dp = [False] * (n + 1)
        dp[0] = True
        for j in range(1, n + 1):
            dp[j] = dp[j - 1] and s2[j - 1] == s3[j - 1]
        for i in range(1, m + 1):
            dp[0] = dp[0] and s1[i - 1] == s3[i - 1]
            for j in range(1, n + 1):
                dp[j] = (dp[j] and s1[i - 1] == s3[i + j - 1]) or \
                        (dp[j - 1] and s2[j - 1] == s3[i + j - 1])
        return dp[n]


if __name__ == "__main__":
    sol = Solution()
    assert sol.isInterleave("aabcc", "dbbca", "aadbbcbcac") is True
    assert sol.isInterleave("aabcc", "dbbca", "aadbbbaccc") is False
    assert sol.isInterleave("", "", "") is True
    assert sol.isInterleave("", "abc", "abc") is True
    assert sol.isInterleave("a", "", "a") is True
    assert sol.isInterleave("a", "b", "ab") is True
    assert sol.isInterleave("a", "b", "ba") is True
    print("All tests passed!")
